File: releasing/commands/src/publish/batchPublish.ts
import { createHash } from 'node:crypto'
import fs from 'node:fs/promises'
import path from 'node:path'

import type { Config, ConfigContext } from '@pnpm/config.reader'
import { PnpmError } from '@pnpm/error'
import type { RunLifecycleHookOptions } from '@pnpm/exec.lifecycle'
import { globalInfo, globalWarn } from '@pnpm/logger'
import { type WebAuthFetchOptions, withOtpHandling } from '@pnpm/network.web-auth'
import type { ExportedManifest } from '@pnpm/releasing.exportable-manifest'
import type { Project } from '@pnpm/types'
import { rimraf } from '@zkochan/rimraf'
import npmFetch from 'npm-registry-fetch'
import { realpathMissing } from 'realpath-missing'
import semver from 'semver'
import { temporaryDirectory } from 'tempy'

import { createPublishSummary, type PublishSummary } from '../tarball/publishSummary.js'
import * as pack from './pack.js'
import { runScriptsIfPresent } from './publish.js'
import {
  createPublishContext,
  createPublishOptions,
  findRegistryInfo,
  isPublishAccess,
} from './publishPackedPkg.js'
import type { PublishRecursiveOpts } from './recursivePublish.js'

export const BATCH_PUBLISH_ENDPOINT = '/-/pnpm/v1/publish'

export type BatchPublishOptions = PublishRecursiveOpts
& Pick<Config, 'embedReadme' | 'ignoreScripts' | 'nodeLinker' | 'packGzipLevel' | 'skipManifestObfuscation'>
& Partial<Pick<ConfigContext, 'hooks'>>

interface PackedPkg {
  project: Project
  publishedManifest: ExportedManifest
  tarballData: Buffer
  summary: PublishSummary
}

/**
 * Publish all {@link pkgs} by sending a single `PUT /-/pnpm/v1/publish` request per target
 * registry, instead of one request per package. The endpoint is not part of the standard npm
 * registry API — the registry has to implement it explicitly (pnpr does), so this whole code
 * path is opt-in via the `--batch` flag.
 *
 * Every package is packed (and its `prepublishOnly`/`prepublish` scripts run) before anything
 * is sent, so a package that fails to pack aborts the publish before any package reaches the
 * registry. The `publish`/`postpublish` scripts run only after the requests succeeded.
 *
 * @param pkgs packages to publish, already filtered (no private packages, no already-published
 *   versions) and in dependency order.
 */
export async function batchPublishPackages (pkgs: Project[], opts: BatchPublishOptions): Promise<PublishSummary[]> {
  if (opts.stage) {
    throw new PnpmError('BATCH_PUBLISH_NO_STAGE', 'Staged publishing cannot be combined with --batch')
  }
  if (opts.provenance) {
    throw new PnpmError('BATCH_PUBLISH_NO_PROVENANCE', 'Provenance statements cannot be generated when publishing with --batch', {
      hint: 'Provenance is bound to a single package, but --batch sends many packages in one request. Publish without --batch to attach provenance.',
    })
  }
  const packedByRegistry = new Map<string, PackedPkg[]>()
  const packedPkgs: PackedPkg[] = []
  for (const project of pkgs) {
    // eslint-disable-next-line no-await-in-loop
    const packedPkg = await packPkgForBatch(project, opts)
    const publishConfigRegistry = typeof packedPkg.publishedManifest.publishConfig?.registry === 'string'
      ? packedPkg.publishedManifest.publishConfig.registry
      : undefined
    const { registry } = findRegistryInfo(packedPkg.publishedManifest, opts, publishConfigRegistry)
    let group = packedByRegistry.get(registry!)
    if (!group) {
      group = []
      packedByRegistry.set(registry!, group)
    }
    group.push(packedPkg)
    packedPkgs.push(packedPkg)
  }
  for (const [registry, group] of packedByRegistry.entries()) {
    for (const { summary } of group) {
      globalInfo(`📦 ${summary.id} → ${registry}`)
    }
    if (opts.dryRun) {
      globalWarn(`Skip publishing ${group.length} package(s) to ${registry} (dry run)`)
      continue
    }
    // eslint-disable-next-line no-await-in-loop
    await multiPublishToRegistry(registry, group, opts)
    globalInfo(`✅ Published ${group.length} package(s) to ${registry} in a single request`)
  }
  if (!opts.ignoreScripts) {
    for (const { project } of packedPkgs) {
      // eslint-disable-next-line no-await-in-loop
      await runScriptsIfPresent(await lifecycleOpts(project.rootDir, opts), ['publish', 'postpublish'], project.manifest)
    }
  }
  return packedPkgs.map(({ summary }) => summary)
}

async function packPkgForBatch (project: Project, opts: BatchPublishOptions): Promise<PackedPkg> {
  if (!opts.ignoreScripts) {
    await runScriptsIfPresent(await lifecycleOpts(project.rootDir, opts), ['prepublishOnly', 'prepublish'], project.manifest)
  }
  // The tarball is packed into a temporary directory and read into memory right away — the
  // request body carries it base64-encoded, so nothing needs to stay on disk.
  const packDestination = temporaryDirectory()
  try {
    const packResult = await pack.api({
      ...opts,
      dir: project.rootDir,
      packDestination,
      dryRun: false,
    })
    const tarballData = await fs.readFile(packResult.tarballPath)
    return {
      project,
      publishedManifest: packResult.publishedManifest,
      tarballData,
      summary: createPublishSummary(packResult, tarballData),
    }
  } finally {
    await rimraf(packDestination)
  }
}

async function lifecycleOpts (pkgRoot: string, opts: BatchPublishOptions): Promise<RunLifecycleHookOptions> {
  return {
    depPath: pkgRoot,
    extraBinPaths: opts.extraBinPaths,
    extraEnv: opts.extraEnv,
    pkgRoot,
    rootModulesDir: await realpathMissing(path.join(pkgRoot, 'node_modules')),
    stdio: 'inherit',
    unsafePerm: true, // when running scripts explicitly, assume that they're trusted.
    userAgent: opts.userAgent,
  }
}

async function multiPublishToRegistry (registry: string, group: PackedPkg[], opts: BatchPublishOptions): Promise<void> {
  // A package-scoped OIDC token cannot authorize a request that publishes many packages, so the
  // OIDC exchange is skipped and only statically configured credentials are used.
  const publishOptions = await createPublishOptions(group[0].publishedManifest, opts, { oidc: false })
  const body = {
    packages: group.map((packedPkg) => createPublishDocument(packedPkg, registry, opts)),
  }
  const fetchOptions: WebAuthFetchOptions = {
    method: 'GET',
    retry: {
      factor: opts.fetchRetryFactor,
      maxTimeout: opts.fetchRetryMaxtimeout,
      minTimeout: opts.fetchRetryMintimeout,
      retries: opts.fetchRetries,
    },
    timeout: opts.fetchTimeout,
  }
  try {
    await withOtpHandling({
      context: createPublishContext(opts),
      fetchOptions,
      operation: async (otp) => npmFetch(BATCH_PUBLISH_ENDPOINT, {
        ...publishOptions,
        access: publishOptions.access ?? undefined,
        otp,
        method: 'PUT',
        body,
        ignoreBody: true,
      } as npmFetch.FetchOptions),
    })
  } catch (err) {
    const statusCode = (err as { statusCode?: number }).statusCode
    if (statusCode === 404 || statusCode === 405) {
      throw new PnpmError(
        'BATCH_PUBLISH_UNSUPPORTED',
        `The registry at ${registry} does not support publishing multiple packages in a single request`,
        {
          hint: `Retry without the --batch flag, or publish to a registry that implements "PUT ${BATCH_PUBLISH_ENDPOINT}" (for example, pnpr).`,
        }
      )
    }
    throw err
  }
}

interface PublishDocument {
  _id: string
  name: string
  description?: string
  'dist-tags': Record<string, string>
  versions: Record<string, ExportedManifest>
  access: 'public' | 'restricted' | null
  _attachments: Record<string, {
    content_type: string
    data: string
    length: number
  }>
}

/**
 * Build the npm publish document for one package — the same JSON body `libnpmpublish` would
 * send as the whole `PUT /:pkg` request. The batch publish request body is an array of these,
 * which lets the registry reuse its single-package publish pipeline per entry.
 */
function createPublishDocument (
  { publishedManifest, tarballData }: Pick<PackedPkg, 'publishedManifest' | 'tarballData'>,
  registry: string,
  opts: Pick<BatchPublishOptions, 'access' | 'tag'>
): PublishDocument {
  const name = publishedManifest.name as string
  const version = semver.clean(publishedManifest.version as string)
  if (!version) {
    throw new PnpmError('BAD_SEMVER', `Invalid semver: ${publishedManifest.version as string}`)
  }
  const publishConfigAccess = publishedManifest.publishConfig?.access
  const access = opts.access ?? (isPublishAccess(publishConfigAccess) ? publishConfigAccess : null)
  if (!name.startsWith('@') && access === 'restricted') {
    throw new PnpmError('UNSCOPED_RESTRICTED', `Can't restrict access to the unscoped package ${name}`)
  }
  const tarballName = `${name}-${version}.tgz`
  const manifest: ExportedManifest = {
    ...publishedManifest,
    _id: `${name}@${version}`,
    version,
    _nodeVersion: process.versions.node,
    dist: {
      integrity: `sha512-${createHash('sha512').update(tarballData).digest('base64')}`,
      shasum: createHash('sha1').update(tarballData).digest('hex'),
      // libnpmpublish stores an http:// URL on purpose (clients fetch via HTTPS regardless when
      // the registry is HTTPS); keep the wire format identical to a single-package publish.
      tarball: new URL(`${name}/-/${tarballName}`, registry).href.replace(/^https:\/\//, 'http://'),
    },
  } as ExportedManifest
  return {
    _id: name,
    name,
    description: publishedManifest.description,
    'dist-tags': { [opts.tag ?? 'latest']: version },
    versions: { [version]: manifest },
    access,
    _attachments: {
      [tarballName]: {
        content_type: 'application/octet-stream',
        data: tarballData.toString('base64'),
        length: tarballData.length,
      },
    },
  }
}

File: pnpm/crates/publish/src/batch_publish.rs
//! Publish several packed packages through pnpr's atomic batch endpoint.

use pnpm_diagnostics::miette::{self, Diagnostic};
use pnpm_network_web_auth::{Host as WebAuthHost, WithOtpError};
use pnpm_reporter::Reporter;
use serde_json::Value;

use crate::{
    failed_to_publish_error::FailedToPublishError,
    global_log::{global_info, global_warn},
    publish_options::{
        PublishUnsupportedRegistryProtocolError, find_registry_info, resolve_access,
    },
    publish_packed_pkg::{
        DistHashes, PackedPkg, PublishHttpError, PublishNetwork, PublishPackedPkgError,
        PublishPackedPkgOptions, build_publish_document, join_registry, publish_with_otp_handling,
        registry_for_display, web_auth_fetch_options,
    },
    publish_summary::{PackedPkgInfo, PublishSummary, create_publish_summary},
    registry_config_keys::NormalizedRegistryUrl,
};

const BATCH_PUBLISH_ENDPOINT: &str = "-/pnpm/v1/publish";

struct BatchGroup {
    registry: NormalizedRegistryUrl,
    package_names: Vec<String>,
    summary_indexes: Vec<usize>,
    documents: Vec<Value>,
}

/// Reject publish modes whose per-package artifacts cannot be represented by
/// one batch request.
pub fn validate_batch_publish_options(
    opts: &PublishPackedPkgOptions,
) -> Result<(), BatchPublishError> {
    if opts.stage {
        return Err(BatchPublishError::Stage);
    }
    if opts.provenance == Some(true) {
        return Err(BatchPublishError::Provenance);
    }
    Ok(())
}

/// Publish every packed package in one request per target registry, calling
/// `on_group_complete` with the input indexes after each completed registry
/// group (including dry-run groups).
pub async fn batch_publish_packed_pkgs<Reporter, OnGroupComplete, Error>(
    packages: &[PackedPkg<'_>],
    opts: &PublishPackedPkgOptions,
    network: &PublishNetwork<'_>,
    mut on_group_complete: OnGroupComplete,
) -> Result<Vec<PublishSummary>, Error>
where
    Reporter: self::Reporter,
    OnGroupComplete: FnMut(&[usize]) -> Result<(), Error>,
    Error: From<BatchPublishError>,
{
    validate_batch_publish_options(opts)?;

    let mut summaries = Vec::with_capacity(packages.len());
    let mut groups: Vec<BatchGroup> = Vec::new();
    for package in packages {
        let manifest = package.published_manifest;
        let name = manifest.get("name").and_then(Value::as_str).unwrap_or_default();
        let publish_config_registry = manifest
            .get("publishConfig")
            .and_then(|config| config.get("registry"))
            .and_then(Value::as_str);
        let registry = find_registry_info(
            name,
            &opts.default_registry,
            &opts.scoped_registries,
            publish_config_registry,
        )
        .map_err(BatchPublishError::from)?;
        let summary = create_publish_summary(
            &PackedPkgInfo {
                published_manifest: manifest,
                tarball_path: package.tarball_path,
                contents: package.contents,
                unpacked_size: package.unpacked_size,
            },
            package.tarball_data,
        );
        let document = build_publish_document(
            manifest,
            package.tarball_data,
            &registry,
            resolve_access(opts.access, manifest),
            &opts.tag,
            &DistHashes { integrity: &summary.integrity, shasum: &summary.shasum },
        )
        .map_err(BatchPublishError::from)?;
        let summary_index = summaries.len();
        summaries.push(summary);

        if let Some(group) = groups.iter_mut().find(|group| group.registry == registry) {
            group.package_names.push(name.to_string());
            group.summary_indexes.push(summary_index);
            group.documents.push(document);
        } else {
            groups.push(BatchGroup {
                registry,
                package_names: vec![name.to_string()],
                summary_indexes: vec![summary_index],
                documents: vec![document],
            });
        }
    }

    let authorizations = if opts.dry_run {
        vec![None; groups.len()]
    } else {
        groups
            .iter()
            .map(|group| batch_authorization(group, network))
            .collect::<Result<Vec<_>, _>>()?
    };

    for (group, authorization) in groups.into_iter().zip(authorizations) {
        let registry = registry_for_display(&group.registry);
        for &summary_index in &group.summary_indexes {
            global_info::<Reporter>(&format!("📦 {} → {registry}", summaries[summary_index].id));
        }
        if opts.dry_run {
            global_warn::<Reporter>(&format!(
                "Skip publishing {} package(s) to {registry} (dry run)",
                group.documents.len(),
            ));
        } else {
            let put_url = join_registry(&group.registry, BATCH_PUBLISH_ENDPOINT)
                .map_err(BatchPublishError::from)?;
            let body = bytes::Bytes::from(
                serde_json::to_vec(&serde_json::json!({ "packages": group.documents }))
                    .expect("serialize batch publish documents"),
            );
            let response = publish_with_otp_handling::<WebAuthHost, Reporter>(
                network.client,
                &put_url,
                authorization.as_deref(),
                "publish",
                body,
                opts.otp.as_deref(),
                false,
                web_auth_fetch_options(&opts.http),
            )
            .await
            .map_err(BatchPublishError::from)?;
            if !response.ok {
                if matches!(response.status, 404 | 405) {
                    return Err(BatchPublishError::Unsupported { registry }.into());
                }
                return Err(BatchPublishError::Failed(FailedToPublishError::new_batch(
                    group.package_names.len(),
                    &registry,
                    response.status,
                    response.status_text,
                    response.body,
                ))
                .into());
            }
            global_info::<Reporter>(&format!(
                "✅ Published {} package(s) to {registry} in a single request",
                group.summary_indexes.len(),
            ));
        }
        on_group_complete(&group.summary_indexes)?;
    }

    Ok(summaries)
}

fn batch_authorization(
    group: &BatchGroup,
    network: &PublishNetwork<'_>,
) -> Result<Option<String>, BatchPublishError> {
    let mut package_names = group.package_names.iter();
    let authorization = package_names.next().and_then(|name| {
        network.auth_headers.for_url_with_package(group.registry.as_str(), Some(name))
    });
    if package_names.any(|name| {
        network.auth_headers.for_url_with_package(group.registry.as_str(), Some(name))
            != authorization
    }) {
        return Err(BatchPublishError::ConflictingCredentials {
            registry: registry_for_display(&group.registry),
        });
    }
    Ok(authorization)
}

/// Failures specific to batch publishing.
#[derive(Debug, derive_more::Display, derive_more::Error, Diagnostic)]
pub enum BatchPublishError {
    #[display("Staged publishing cannot be combined with --batch")]
    #[diagnostic(code(ERR_PNPM_BATCH_PUBLISH_NO_STAGE))]
    Stage,

    #[display("Provenance statements cannot be generated when publishing with --batch")]
    #[diagnostic(
        code(ERR_PNPM_BATCH_PUBLISH_NO_PROVENANCE),
        help(
            "Provenance is bound to a single package, but --batch sends many packages in one request. Publish without --batch to attach provenance."
        )
    )]
    Provenance,

    #[display(
        "Packages targeting {registry} resolve to different authentication credentials and cannot be published in one batch"
    )]
    #[diagnostic(
        code(ERR_PNPM_BATCH_PUBLISH_CONFLICTING_CREDENTIALS),
        help(
            "Configure one credential that can publish every package targeting this registry, or publish without --batch."
        )
    )]
    ConflictingCredentials {
        #[error(not(source))]
        registry: String,
    },

    #[display(
        "The registry at {registry} does not support publishing multiple packages in a single request"
    )]
    #[diagnostic(
        code(ERR_PNPM_BATCH_PUBLISH_UNSUPPORTED),
        help(
            r#"Retry without the --batch flag, or publish to a registry that implements "PUT /-/pnpm/v1/publish" (for example, pnpr)."#
        )
    )]
    Unsupported {
        #[error(not(source))]
        registry: String,
    },

    #[display("{_0}")]
    #[diagnostic(transparent)]
    Registry(PublishUnsupportedRegistryProtocolError),

    #[display("{_0}")]
    #[diagnostic(transparent)]
    Package(PublishPackedPkgError),

    #[display("{_0}")]
    #[diagnostic(transparent)]
    Otp(WithOtpError<PublishHttpError>),

    #[display("{_0}")]
    #[diagnostic(transparent)]
    Failed(FailedToPublishError),
}

impl From<PublishUnsupportedRegistryProtocolError> for BatchPublishError {
    fn from(error: PublishUnsupportedRegistryProtocolError) -> Self {
        BatchPublishError::Registry(error)
    }
}

impl From<PublishPackedPkgError> for BatchPublishError {
    fn from(error: PublishPackedPkgError) -> Self {
        BatchPublishError::Package(error)
    }
}

impl From<WithOtpError<PublishHttpError>> for BatchPublishError {
    fn from(error: WithOtpError<PublishHttpError>) -> Self {
        BatchPublishError::Otp(error)
    }
}
